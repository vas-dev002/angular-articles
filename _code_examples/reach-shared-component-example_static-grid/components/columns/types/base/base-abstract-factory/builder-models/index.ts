export * from "./datetime-filtering-option";
export * from "./i-cell-class-options";
export * from "./i-datetime-filter-options";
export * from "./i-datetime-filter-params";
export * from "./i-filtering-options";
export * from "./i-grouping-options";
export * from "./i-sorting-options";
export * from "./i-text-filter-options";
export * from "./i-text-filter-params";
export * from "./i-text-floating-filter-component-params";
export * from "./i-tooltip-by-field-options";
export * from "./i-tooltip-by-function-options";
export * from "./i-tooltip-options";
export * from "./i-value-formatter-options";
export * from "./i-value-getter-by-function-options";
export * from "./i-value-getter-by-property-options";
export * from "./i-value-setter-options";
export * from "./i-visibility-options";
export * from "./i-width-options";
export * from "./menu-tab";
export * from "./sorting-order-option";
export * from "./text-filtering-option";
