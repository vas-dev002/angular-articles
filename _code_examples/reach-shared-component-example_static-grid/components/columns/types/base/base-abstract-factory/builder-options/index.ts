export * from "./cell-class-options";
export * from "./sorting-options";
export * from "./text-filter-options";
export * from "./tooltip-by-function-options";
export * from "./visibility-options";
export * from "./width-options";
