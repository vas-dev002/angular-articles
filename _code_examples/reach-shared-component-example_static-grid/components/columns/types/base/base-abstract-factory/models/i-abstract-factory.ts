import { ColDef } from "@ag-grid-community/core";
import { ITypeControllerParams } from "../../../../../../../../../../components/shared/aggrid/static-grid";

export interface IColDefAbstractFactory {
	createColDef: (controllerParams: ITypeControllerParams) => ColDef;
}
