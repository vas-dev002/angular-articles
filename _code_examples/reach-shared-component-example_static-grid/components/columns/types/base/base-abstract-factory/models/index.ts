export * from "./i-abstract-factory";
