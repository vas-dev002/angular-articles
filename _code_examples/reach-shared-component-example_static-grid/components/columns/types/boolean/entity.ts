export interface BooleanGroupParams {
	value: string;
}
