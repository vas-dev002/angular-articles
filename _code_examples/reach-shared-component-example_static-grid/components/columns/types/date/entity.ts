import { ValueFormatterParams } from "@ag-grid-community/core";

export interface IDateValue {
	date: string;
}

export interface IDateGroupKeyParams {
	value: string;
	date: string;
}

export interface IDateValueFormatterParams extends ValueFormatterParams {
	value: string | IDateValue;
}
