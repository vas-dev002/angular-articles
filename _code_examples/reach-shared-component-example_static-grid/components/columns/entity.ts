import { Type, Injector } from "@angular/core";
import { ICellRendererAngularComp } from "@ag-grid-community/angular";
import { ColDef } from "@ag-grid-community/core";
import { IColumnMetaSG } from "../main/entity";
import { StaticGridComponent } from "../main/comp";
import { ColumnDisplayType } from "./types";
import { IColDefAbstractFactory } from "./types/base/base-abstract-factory/models";
import { ObjectDetailsType } from "@evergreen-shared/entity";

export interface ITypeControllerSG {
	renderComponent: Type<ICellRendererAngularComp>;
	columnDefinition(): ColDef;
}

export interface IFilterOptionsSG {
	text: string;
	value: unknown;
	props?: Object;
	sortOrder?: number;
}

export interface ITypeControllerParams {
	meta: IColumnMetaSG;
	grid: StaticGridComponent;
	customOptions: Object;
	injector?: Injector;
}

export interface ILinkMeta {
	uRLColumnName: string;
	objectType: string;
	objectTypeColumnName: string;
	labelColumnName: string;
	keyColumnName: string;
	parameters: string;
	urlParameters: string[];
	renderLinkProperty: string;
}

export interface IOutcomeComparatorValue {
	status: number;
	text: string;
	tooltip: string;
}

export enum ApplicationForwardPathTypeEnum {
	None = 0,
	Retire = 1,
	Keep = 2,
	ForwardPath = 3,
	Uncategorized = 4
}

export enum ApplicationAssociationTypeEnum {
	None = 0,
	Dynamic = 1,
	StickyInclude = 2,
	StickyExclude = 3,
	ForwardPathed = 4,
	Mixed = 5
}

export enum StringValueGetterMode {
	String,
	SelectedItem,
	Params
}

export interface ITypedController {
	key: ColumnDisplayType;
	factory: Type<IColDefAbstractFactory>;
}

export enum TaskFlags {
	IsBucketTask = "IsBucketTask",
	IsReadOnlyTask = "IsReadOnlyTask",
	IsSystemTask = "IsSystemTask"
}

export interface IDateTimeEditorChanges {
	taskId: number;
	projectId: number;
	id: number;
	objectType: ObjectDetailsType;
	date: string;
	postBackParameters: { [key: string]: unknown };
}

export interface ICapacityDateTimeEditorChanges extends IDateTimeEditorChanges {
	slotId: number;
	slotName: string;
}
