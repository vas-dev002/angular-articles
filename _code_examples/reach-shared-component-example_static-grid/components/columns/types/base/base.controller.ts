import { IColumnMetaSG, IGetCellClassParams, FilterTypeSG } from "../../../main/entity";
import { Type, Injector } from "@angular/core";
import { ICellRendererAngularComp } from "@ag-grid-community/angular";
import { ColDef, ITooltipParams, ValueFormatterParams } from "@ag-grid-community/core";
import { ITypeControllerParams, IFilterOptionsSG } from "../../entity";
import { StaticGridComponent } from "../../../main/comp";
import { isColumnDisabled } from "../../../main/config/main";
import { uniqBy, orderBy } from "lodash";
import { StringMultiSelectFilterComponent } from "../../../../../filter/stringMultiSelect/comp";
import { MultiselectFloatingFilterComponent } from "../../../../../filter/floatingMultiselect/comp";
import { TextFloatingFilterComponent } from "../../../../../filter/floatingTextInput/comp";
import { customTextComparator, valueToString } from "../../../../../util/column";
import { MaterialTooltipComponent } from "../../../../../../material-tooltip/material-tooltip.component";
import { isNullOrUndefined } from "../../../../../../../util/common/common";

/**
 * @deprecated. Use builder and factory instead.
 */
export abstract class BaseTypeController {

	public abstract renderComponent: Type<ICellRendererAngularComp>;
	public customFilterFramework: any = StringMultiSelectFilterComponent;
	public meta: IColumnMetaSG;
	public grid: StaticGridComponent;
	public injector: Injector;

	constructor(protected readonly controllerParams: ITypeControllerParams) {
		this.meta = controllerParams.meta;
		this.grid = controllerParams.grid;
		this.injector = controllerParams.injector;
	}

	protected get baseDef(): ColDef {
		return {
			cellRendererFramework: this.renderComponent,
			field: this.meta.columnName,
			cellClass: this.meta.cssClass,
			headerName: this.meta.headerName,
			hide: !this.meta.visible,
			menuTabs: ["generalMenuTab", "filterMenuTab", "columnsMenuTab"],
			minWidth: 50,
			tooltipField: this.meta.tooltipField,
			maxWidth: 500,
			lockVisible: this.meta.lockVisible,
			keyCreator: this.getGroupKey(),
			cellClassRules: this.getCssRules(this.meta),
			tooltipComponentFramework: MaterialTooltipComponent,
			tooltipValueGetter: this.getTooltip(),
			valueFormatter: this.getStringValue(),
			...this.customFilterDef(),
			...this.addHeaderTooltip(),
			...this.customSortDef
		};
	}

	protected customFilterDef(): ColDef {
		return this.getFilterDefObject(this.meta.filterType);
	}

	protected get customSortDef(): ColDef {
		return {
			sortingOrder: ["asc", "desc", null]
		};
	}

	protected getStringValue(property?: string): (params: ValueFormatterParams) => string {
		return (params: ValueFormatterParams) => {
			return valueToString(params.value, property);
		};
	}

	protected getGroupKey(): <T>(params: { value: T }) => string {
		return <T>(params: { value: T }) => {
			return JSON.stringify(params.value);
		};
	}

	protected getTooltip(property?: string): (params: ITooltipParams) => string {
		return (params: ITooltipParams) => {
			return valueToString(params.value, property);
		};
	}

	protected getCssRules(meta: IColumnMetaSG) {
		return {
			"sg-cell-disabled": (params: IGetCellClassParams<{ [disabledField: string]: boolean }, null>) => {
				return isColumnDisabled(params.data, meta);
			}
		};
	}

	protected getMetaWithFilterOptions() {
		const meta = this.meta.metaFromApi || {};
		return {
			...meta,
			filterOptions: meta["filterOptions"] || this.getFilterOptions()
		};
	}

	protected customSortFilterOptions(filterOptions: IFilterOptionsSG[]): IFilterOptionsSG[] {
		return orderBy(
			filterOptions,
			[(option) => option.text.toString().toLowerCase()],
			["asc"]
		).sort((a, b) => { // "Empty" to the top.
			const comparator = (val) => val.text === "Empty" ? 0 : 1;
			return comparator(a) - comparator(b);
		});
	}

	protected getCustomModifiedOptionValue(value): IFilterOptionsSG {
		return {
			text: value || "Empty",
			value: value || ""
		};
	}

	// Remove for release! It's for testers and devs only.
	private addHeaderTooltip() {
		return this.meta.metaFromApi && !isNullOrUndefined(this.meta.metaFromApi["customUiTransformation"])
			? { headerTooltip: this.meta.metaFromApi["customUiTransformation"] || "String" }
			: {};
	}

	getFilterOptions(): IFilterOptionsSG[] {
		const uniqValues = uniqBy(this.grid.rowData, this.meta.columnName);
		const options = [];
		uniqValues.forEach(value => {
			options.push(this.getCustomModifiedOptionValue(value[this.meta.columnName]));
		});
		return this.customSortFilterOptions(options);
	}

	// Need to add specific interfaces for this objects (instead of ColDef type).
	// And probably it will be better to put this into the utilities file.
	getFilterDefObject(filterType: FilterTypeSG): ColDef {
		switch (filterType) {
			case FilterTypeSG.None:
				return {
					filter: false,
					filterParams: {}
				} as ColDef;
			case FilterTypeSG.Multiselect:
				return {
					filterFramework: this.customFilterFramework,
					floatingFilterComponentFramework: MultiselectFloatingFilterComponent,
					floatingFilterComponentParams: {
						suppressFilterButton: true,
						meta: this.getMetaWithFilterOptions()
					},
					filterParams: {
						defaultOption: "equals",
						filterOptions: ["equals", "notEquals"],
						textFormatter: function (r) { return r.toLowerCase(); },
						selectAllOnMiniFilter: false,
						meta: this.getMetaWithFilterOptions()
					}
				} as ColDef;
			case FilterTypeSG.Number:
			case FilterTypeSG.String:
			case FilterTypeSG.Date:
			case FilterTypeSG.DateTime:
			default:
				return {
					filter: "agTextColumnFilter",
					floatingFilterComponentFramework: TextFloatingFilterComponent,
					floatingFilterComponentParams: {
						suppressFilterButton: true
					},
					filterParams: {
						defaultOption: "contains",
						filterOptions: [
							"contains",
							"notContains",
							"startsWith",
							"endsWith"
						],
						caseSensitive: true,
						textCustomComparator: customTextComparator
					}
				} as ColDef;
		}
	}

}
