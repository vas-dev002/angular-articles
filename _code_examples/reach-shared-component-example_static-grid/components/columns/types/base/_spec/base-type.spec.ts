import { Type } from "@angular/core";
import { ICellRendererAngularComp } from "@ag-grid-community/angular";
import { IColumnMetaSG, FilterTypeSG, IGetCellClassParams, IDataRowSG } from "../../../../main/entity";
import { StaticGridComponent } from "../../../../main/comp";
import { ColDef, ITooltipParams, ValueFormatterParams } from "@ag-grid-community/core";
import { BaseTypeController } from "../base.controller";
import { getValueGetterResults } from "./types-test.utils";
import { IColDefAbstractFactory } from "../base-abstract-factory/models";

declare const expect: jest.Expect;

class TestTypeController extends BaseTypeController {
	public renderComponent = {} as Type<ICellRendererAngularComp>;
	public columnDefinition(): ColDef {
		return {
			...this.baseDef
		};
	}
}

describe("Static Grid Column Type", () => {
	describe(`Base Controller: BaseControllerSG`, () => {
		const baseMeta: IColumnMetaSG = {
			headerName: "Test Header",
			visible: false,
			cssClass: "testCss",
			columnName: "testColumn",
			controller: {} as Type<IColDefAbstractFactory>,
			filterType: FilterTypeSG.Multiselect,
			lockVisible: false,
			isObjectKey: false,
			metaFromApi: {}
		};
		let meta;
		let customOptions = {};
		let grid = { rowData: [] } as StaticGridComponent;
		let controller: TestTypeController;

		beforeEach(() => {
			meta = { ...baseMeta };
			customOptions = {};
			grid = { rowData: [] } as StaticGridComponent;
			controller = new TestTypeController({ meta, grid, customOptions });
		});

		it("should return default column definition", () => {
			const definition = controller.columnDefinition();
			expect(definition).toMatchSnapshot();
		});

		it("should return default column definition with None filter type", () => {
			meta.filterType = FilterTypeSG.None;
			controller = new TestTypeController({ meta, grid, customOptions });
			const definition = controller.columnDefinition();
			expect(definition).toMatchSnapshot();
		});

		it("should return keyCreator for groups", () => {
			meta = {} as IColumnMetaSG;
			controller = new TestTypeController({ meta, grid, customOptions });
			const definition = controller.columnDefinition();
			expect(definition.keyCreator).toBeDefined();
			const values = [1, "test", { test: "value" }];
			const res = {};
			values.forEach((value) => {
				res[value.toString()] = definition.keyCreator({ value });
			});
			expect(res).toMatchSnapshot();
		});

		it("should fill tooltipField", () => {
			const column = "TooltipTestColumn";
			meta = { tooltipField: column } as IColumnMetaSG;
			controller = new TestTypeController({ meta, grid, customOptions });
			const definition = controller.columnDefinition();
			expect(definition.tooltipField).toBeDefined();
			expect(definition.tooltipField).toBe(column);
		});

		it("should return tooltip by default", () => {
			meta = {} as IColumnMetaSG;
			controller = new TestTypeController({ meta, grid, customOptions });
			const definition = controller.columnDefinition();
			expect(definition.tooltipValueGetter).toBeDefined();
			const params = { value: "TestTooltip" } as ITooltipParams;
			expect(definition.tooltipValueGetter(params)).toBe(params.value);
		});

		it("should return export value by default", () => {
			meta = {} as IColumnMetaSG;
			controller = new TestTypeController({ meta, grid, customOptions });
			const definition = controller.columnDefinition();
			expect(definition.valueFormatter).toBeDefined();
			const params = { value: "TestValue" } as ValueFormatterParams;
			if (typeof definition.valueFormatter !== "string") {
				expect(definition.valueFormatter(params)).toBe(params.value);
			} else {
				expect(definition.valueFormatter).toBe(params.value);
			}
		});

		it("should set class by disabled status", () => {
			const column = "DisabledTestColumn";
			meta = { disabledField: column } as IColumnMetaSG;
			controller = new TestTypeController({ meta, grid, customOptions });
			const definition = controller.columnDefinition();
			expect(definition.cellClassRules).toBeDefined();
			const cellClass = "sg-cell-disabled";
			const cssFunction = definition.cellClassRules[cellClass] as Function;
			expect(typeof cssFunction).toBe("function");
			let params = {
				data: null,
				rowIndex: 0, node: null, value: null, api: null
			} as IGetCellClassParams<{ [disabledField: string]: boolean }, null>;
			expect(cssFunction(params)).toBe(false);
			const values = [true, false, null, undefined];
			const results = [true, false, false, false];
			values.forEach((value, index) => {
				params = {
					data: { [column]: value },
					rowIndex: 0, node: null, value: null, api: null
				} as IGetCellClassParams<{ [disabledField: string]: boolean }, null>;
				expect(cssFunction(params)).toBe(results[index]);
			});

		});

		it("should return the value", () => {
			meta = {} as IColumnMetaSG;
			controller = new TestTypeController({ meta, grid, customOptions });
			const definition = controller.columnDefinition();
			const valueFunction = definition.valueFormatter;
			const value = "ShowMe";
			const params = { value } as ValueFormatterParams;
			if (typeof valueFunction !== "string") {
				expect(valueFunction(params)).toBe(value);
			} else {
				expect(valueFunction).toBe(value);
			}
		});

		it("should return default column definition with Number filterType", () => {
			const metaWithNumber = { ...meta, filterType: FilterTypeSG.Number };
			const newController = new TestTypeController({ meta: metaWithNumber, grid, customOptions });
			const definition = newController.columnDefinition();
			expect(definition).toMatchSnapshot();
		});

		it("should handle Empty values when generating FilterOptions", () => {
			const gridWithRowData = {
				rowData: [
					{
						testColumn: ""
					} as IDataRowSG,
					{
						testColumn: "value"
					} as IDataRowSG
				]
			} as StaticGridComponent;
			const newController = new TestTypeController({ meta, grid: gridWithRowData, customOptions });
			const definition = newController.columnDefinition();
			expect(definition.floatingFilterComponentParams.meta.filterOptions).toMatchSnapshot();
		});

		it("should handle customUiTransformation value in metaFromApi", () => {
			const metaWithCustomUiTransformation = {
				...meta,
				metaFromApi: {
					customUiTransformation: ""
				}
			};
			const newController = new TestTypeController({ meta: metaWithCustomUiTransformation, grid, customOptions });
			const definition = newController.columnDefinition();
			expect(definition.filterParams).toMatchSnapshot();
		});

		it("[Common type test function: getValueGetterResults should handle empty value getter]", () => {
			expect(getValueGetterResults("")).toMatchSnapshot();
		});

	});
});
