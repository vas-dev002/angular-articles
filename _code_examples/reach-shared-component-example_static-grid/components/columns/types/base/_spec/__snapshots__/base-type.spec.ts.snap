// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Static Grid Column Type Base Controller: BaseControllerSG [Common type test function: getValueGetterResults should handle empty value getter] 1`] = `Object {}`;

exports[`Static Grid Column Type Base Controller: BaseControllerSG should handle Empty values when generating FilterOptions 1`] = `
Array [
  Object {
    "text": "Empty",
    "value": "",
  },
  Object {
    "text": "value",
    "value": "value",
  },
]
`;

exports[`Static Grid Column Type Base Controller: BaseControllerSG should handle customUiTransformation value in metaFromApi 1`] = `
Object {
  "defaultOption": "equals",
  "filterOptions": Array [
    "equals",
    "notEquals",
  ],
  "meta": Object {
    "customUiTransformation": "",
    "filterOptions": Array [],
  },
  "selectAllOnMiniFilter": false,
  "textFormatter": [Function],
}
`;

exports[`Static Grid Column Type Base Controller: BaseControllerSG should return default column definition 1`] = `
Object {
  "cellClass": "testCss",
  "cellClassRules": Object {
    "sg-cell-disabled": [Function],
  },
  "cellRendererFramework": Object {},
  "field": "testColumn",
  "filterFramework": [Function],
  "filterParams": Object {
    "defaultOption": "equals",
    "filterOptions": Array [
      "equals",
      "notEquals",
    ],
    "meta": Object {
      "filterOptions": Array [],
    },
    "selectAllOnMiniFilter": false,
    "textFormatter": [Function],
  },
  "floatingFilterComponentFramework": [Function],
  "floatingFilterComponentParams": Object {
    "meta": Object {
      "filterOptions": Array [],
    },
    "suppressFilterButton": true,
  },
  "headerName": "Test Header",
  "hide": true,
  "keyCreator": [Function],
  "lockVisible": false,
  "maxWidth": 500,
  "menuTabs": Array [
    "generalMenuTab",
    "filterMenuTab",
    "columnsMenuTab",
  ],
  "minWidth": 50,
  "sortingOrder": Array [
    "asc",
    "desc",
    null,
  ],
  "tooltipComponentFramework": [Function],
  "tooltipField": undefined,
  "tooltipValueGetter": [Function],
  "valueFormatter": [Function],
}
`;

exports[`Static Grid Column Type Base Controller: BaseControllerSG should return default column definition with None filter type 1`] = `
Object {
  "cellClass": "testCss",
  "cellClassRules": Object {
    "sg-cell-disabled": [Function],
  },
  "cellRendererFramework": Object {},
  "field": "testColumn",
  "filter": false,
  "filterParams": Object {},
  "headerName": "Test Header",
  "hide": true,
  "keyCreator": [Function],
  "lockVisible": false,
  "maxWidth": 500,
  "menuTabs": Array [
    "generalMenuTab",
    "filterMenuTab",
    "columnsMenuTab",
  ],
  "minWidth": 50,
  "sortingOrder": Array [
    "asc",
    "desc",
    null,
  ],
  "tooltipComponentFramework": [Function],
  "tooltipField": undefined,
  "tooltipValueGetter": [Function],
  "valueFormatter": [Function],
}
`;

exports[`Static Grid Column Type Base Controller: BaseControllerSG should return default column definition with Number filterType 1`] = `
Object {
  "cellClass": "testCss",
  "cellClassRules": Object {
    "sg-cell-disabled": [Function],
  },
  "cellRendererFramework": Object {},
  "field": "testColumn",
  "filter": "agTextColumnFilter",
  "filterParams": Object {
    "caseSensitive": true,
    "defaultOption": "contains",
    "filterOptions": Array [
      "contains",
      "notContains",
      "startsWith",
      "endsWith",
    ],
    "textCustomComparator": [Function],
  },
  "floatingFilterComponentFramework": [Function],
  "floatingFilterComponentParams": Object {
    "suppressFilterButton": true,
  },
  "headerName": "Test Header",
  "hide": true,
  "keyCreator": [Function],
  "lockVisible": false,
  "maxWidth": 500,
  "menuTabs": Array [
    "generalMenuTab",
    "filterMenuTab",
    "columnsMenuTab",
  ],
  "minWidth": 50,
  "sortingOrder": Array [
    "asc",
    "desc",
    null,
  ],
  "tooltipComponentFramework": [Function],
  "tooltipField": undefined,
  "tooltipValueGetter": [Function],
  "valueFormatter": [Function],
}
`;

exports[`Static Grid Column Type Base Controller: BaseControllerSG should return keyCreator for groups 1`] = `
Object {
  "1": "1",
  "[object Object]": "{\\"test\\":\\"value\\"}",
  "test": "\\"test\\"",
}
`;
